### PR TITLE
TTML python module: add ModuleBase interface abstract base class for composing TTML Modules in python

### PR DESCRIPTION
**Pull request by athompsonTT**
### Ticket
N/A

### Problem description
A standard way to define TTML Modules in Python using `ttml` Python module's primitives is desired.

### What's changed
Introduced a comprehensive module base class system for TTML, enabling both C++ and Python module implementations with seamless interoperability. The implementation provides a PyTorch-like interface.

Added nanogpt examples.

### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes ([latest run](https://github.com/tenstorrent/tt-metal/actions/runs/21079456334))
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

File: tt-train/sources/ttml/nanobind/nb_ops.cpp [hunk 1]
```diff
@@ -131,7 +139,21 @@ void py_module(nb::module_& m) {
 
     {
         auto py_linear = static_cast<nb::module_>(m.attr("linear"));
-        py_linear.def("linear_op", &ttml::ops::linear_op, nb::arg("tensor"), nb::arg("weight"), nb::arg("bias"));
+        // Overload with optional bias (None support)
+        py_linear.def(
+            "linear_op",
+            [](const autograd::TensorPtr& tensor,
+               const autograd::TensorPtr& weight,
+               nb::object bias_obj) -> autograd::TensorPtr {
```
Comment (ayerofieiev-tt) [inline tt-train/sources/ttml/nanobind/nb_ops.cpp]: I think we should discuss what can be done to avoid this
I assume this is all because bias can be a Tensor or a None.
TensorPtr should conceptually support None just find

Comment (nsextonTT) [reply tt-train/sources/ttml/nanobind/nb_ops.cpp]: Yeah, I don't see why this just can't be an `std::optional<autograd::TensorPtr>`?

File: tt-train/sources/examples/linear_regression/linear_regression.py [hunk 2]
```diff
@@ -31,6 +31,7 @@
 # Import TTML (adjust path if needed)
 # ---------------------------------------------------------------------------
 sys.path.append(f"{os.environ['TT_METAL_HOME']}/tt-train/sources/ttml")
```
Comment (abustamanteTT) [inline tt-train/sources/examples/linear_regression/linear_regression.py]: Nit: I don't think this is necessary anymore after the packaging PRs

File: tt-train/sources/examples/nano_gpt/nanogpt_full_cpp_example.py [hunk 3]
```diff
@@ -0,0 +1,1362 @@
...
+import ttml
+from ttml.models.nanogpt import NanoGPT, NanoGPTConfig, create_nanogpt
+from ttml.modules import Parameter
+
+
+@dataclass
+class TrainingConfig:
+    """Training configuration matching C++ TrainingConfig."""
+
+    project_name: str = "tt_train_nano_gpt"
```
Comment (abustamanteTT) [inline tt-train/sources/examples/nano_gpt/nanogpt_full_cpp_example.py]: I'm torn on this; the example does 1-to-1 match the original C++, but we also have TrainingConfig, ModelConfig, and the appropriate parsers in `ttml.common.config`. I feel like while having the exact match is great as a start, perhaps swapping in our existing Python config classes will go a longer way to showing that we have the infrastructure necessary to do this all in Python

File: tt-train/sources/examples/nano_gpt/nanogpt_full_cpp_example.py [hunk 4]
```diff
@@ -0,0 +1,1362 @@
...
+
+def read_file_to_str(file_path: str) -> str:
+    """Read file to string."""
+    with open(file_path, "r", encoding="utf-8") as f:
+        return f.read()
+
+
+def round_up_to_tile(value: int, tile_size: int = 32) -> int:
+    """Round up value to nearest multiple of tile_size."""
+    return ((value + tile_size - 1) // tile_size) * tile_size
```
Comment (abustamanteTT) [inline tt-train/sources/examples/nano_gpt/nanogpt_full_cpp_example.py]: See above comment; we have this in `ttml.common.utils`, I believe

File: tt-train/sources/examples/nano_gpt/nanogpt_full_cpp_example.py (hunk 3 above)
Comment (athompsonTT) [reply tt-train/sources/examples/nano_gpt/nanogpt_full_cpp_example.py]: I'll see what I can do wrt using ttml.common.config

File: tt-train/sources/examples/nano_gpt/nanogpt_full_cpp_example.py [hunk 5]
```diff
@@ -0,0 +1,1362 @@
...
+            if len(seq) == sequence_length and len(target) == sequence_length:
+                dataset.append((seq, target))
+
+        return dataset, tokenizer
+    else:
+        raise ValueError(f"Unsupported tokenizer type: {tokenizer_type}")
+
+
+def create_mask(sequence_length: int, device) -> ttml.autograd.Tensor:
+    """Create causal attention mask matching C++ implementation."""
```
Comment (abustamanteTT) [inline tt-train/sources/examples/nano_gpt/nanogpt_full_cpp_example.py]: `ttml.common.utils`

File: tt-train/sources/examples/nano_gpt/nanogpt_full_cpp_example.py [hunk 6]
```diff
@@ -0,0 +1,1362 @@
...
+    model: NanoGPT,
+    tokenizer: CharTokenizer,
+    prompt: str,
+    max_new_tokens: int,
+    sequence_length: int,
+    mask: ttml.autograd.Tensor,
+    temperature: float = 1.0,
+    top_k: int = 0,
+) -> str:
+    """Generate text from a prompt using sampling with temperature.
```
Comment (abustamanteTT) [inline tt-train/sources/examples/nano_gpt/nanogpt_full_cpp_example.py:390]: Not necessarily a concern for this PR, but I've been thinking for a while now that we have re-implementations of this sampling across multiple examples. We may be better served by following the Hugging Face style of giving the model itself a `generate()` function, or otherwise putting such a function in `ttml.common.utils`; I'm open to thoughts

File: tt-train/sources/ttml/nanobind/nb_ops.cpp [hunk 7]
```diff
@@ -131,7 +140,17 @@ void py_module(nb::module_& m) {
 
     {
         auto py_linear = static_cast<nb::module_>(m.attr("linear"));
-        py_linear.def("linear_op", &ttml::ops::linear_op, nb::arg("tensor"), nb::arg("weight"), nb::arg("bias"));
+        // Overload with optional bias (None support)
+        py_linear.def(
+            "linear_op",
```
Comment (ayerofieiev-tt) [inline tt-train/sources/ttml/nanobind/nb_ops.cpp]: can be ?
```suggestion
            "linear",
```

File: tt-train/sources/examples/nano_gpt/nanogpt_full_cpp_example.py [hunk 8]
```diff
@@ -0,0 +1,1362 @@
...
+    print("=" * 70)
+    print()
+
+    # Set TT_METAL_RUNTIME_ROOT if not set and TT_METAL_HOME is available
+    # This is needed for the runtime to find kernel files like moreh_mean
+    if "TT_METAL_RUNTIME_ROOT" not in os.environ:
+        tt_metal_home = os.environ.get("TT_METAL_HOME", "")
+        if tt_metal_home and os.path.exists(tt_metal_home):
+            os.environ["TT_METAL_RUNTIME_ROOT"] = tt_metal_home
+            print(f"Set TT_METAL_RUNTIME_ROOT={tt_metal_home} (from TT_METAL_HOME)")
```
Comment (abustamanteTT) [inline tt-train/sources/examples/nano_gpt/nanogpt_full_cpp_example.py]: We've got `get_tt_metal_home()` in `ttml.common.utils`, that would reduce the number of os.environ calls here

File: tt-train/sources/ttml/nanobind/__init__.cpp [hunk 9]
```diff
@@ -32,7 +32,11 @@ NB_MODULE(_ttml, m) {
     // Bind NamedParameters as a proper map type at the top level
     nb::bind_map<ttml::serialization::NamedParameters>(m, "NamedParameters");
 
-    ttml::nanobind::util::export_enum<tt::tt_metal::Layout>(m);
+    // NOTE: Layout enum is NOT exported here.
+    // ttnn's C++ library (which _ttml links against) already exports Layout.
+    // Python code should use: from ttnn import Layout
+    // See ttml/__init__.py which imports Layout from ttnn.
```
Comment (abustamanteTT) [inline tt-train/sources/ttml/nanobind/__init__.cpp:42]: Love this note; longer term, I'd love to see details on the Python API documented with explanations like this

**Review by abustamanteTT — APPROVED** [review]
Once the comments are resolved, I think it's good to go. Great job!

File: tt-train/sources/ttml/ops/reshape_op.cpp [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+
+#include <core/ttnn_all_includes.hpp>
+
+#include "autograd/auto_context.hpp"
+#include "autograd/graph_utils.hpp"
+#include "autograd/tensor.hpp"
+
+namespace ttml::ops {
+
+autograd::TensorPtr reshape(const autograd::TensorPtr& tensor, std::span<uint32_t> shape) {
```
Comment (philei-tt) [inline tt-train/sources/ttml/ops/reshape_op.cpp:15]: Why did you need to add this op, and why it was not required before?

File: tt-train/configs/training_configs/training_shakespeare_nanogpt.yaml [hunk 11]
```diff
@@ -13,7 +13,7 @@ training_config:
   clip_grad_norm_max_norm: 1.0
   tokenizer_type: "char"
   data_path: "data/shakespeare.txt"
-  model_config: "configs/model_configs/nanogpt.yaml"
+  model_config: "model_configs/nanogpt.yaml"
```
Comment (philei-tt) [inline tt-train/configs/training_configs/training_shakespeare_nanogpt.yaml]: Why change?

File: tt-train/sources/ttml/ttml/models/nanogpt/embedding.py [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+    def forward(self, x: ttml.autograd.Tensor) -> ttml.autograd.Tensor:
+        """Forward pass of embedding layer.
+
+        Args:
+            x: Input tensor of token indices, shape [batch_size, 1, 1, seq_len]
+
+        Returns:
+            Output tensor of embeddings
+        """
+        return ttml.ops.embedding.embedding_op(x, self.weight.tensor)
```
Comment (philei-tt) [inline tt-train/sources/ttml/ttml/models/nanogpt/embedding.py]: Any reason to keep _op suffix? It's already in ops namespace.

File: tt-train/sources/ttml/ttml/modules/module_base.py [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+
+    def __setattr__(self, name: str, value: Any) -> None:
+        object.__setattr__(self, name, value)
+
+        # Skip if not initialized yet
+        if "_buffers" not in self.__dict__:
+            return
+
+        # Auto-register modules
+        if isinstance(value, CppModuleBase):
```
Comment (philei-tt) [inline tt-train/sources/ttml/ttml/modules/module_base.py:28]: Just to make sure this will also catch modules defined in python, right? Python modules inherit from AbstractModuleBase which inherits from CppModuleBase, right?

File: tt-train/sources/ttml/ttml/models/nanogpt/embedding.py [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+
+    def __call__(self, x: ttml.autograd.Tensor) -> ttml.autograd.Tensor:
+        """Call the forward method."""
+        return self.forward(x)
```
Comment (philei-tt) [inline tt-train/sources/ttml/ttml/models/nanogpt/embedding.py]: Can't this be moved to AbstractModuleBase?

File: tt-train/sources/ttml/ttml/models/nanogpt/__init__.py [hunk 15]
```diff
@@ -0,0 +1,173 @@
...
+    n_embd: int = 768  # Embedding dimension
+    n_layer: int = 12  # Number of transformer blocks
+    n_head: int = 12  # Number of attention heads
+    dropout: float = (
+        0.2  # Dropout probability (matching C++ default: dropout_prob = 0.2F)
+    )
+    bias: bool = True  # Use bias in linear layers and layer norm
+
+
+class NanoGPT(AbstractModuleBase):
```
Comment (philei-tt) [inline tt-train/sources/ttml/ttml/models/nanogpt/__init__.py:140]: ttml.modules.AbstractModuleBase naming is a bit too expressive, no? Why not just ttml.Module or ttml.modules.Module?

**Comment by philei-tt** [comment]
Can you add instructions on how to install ttml python package? How to run model? Docs on how to create modules and build your own model in ttml/tt-train?

File: tt-train/configs/training_configs/training_shakespeare_nanogpt.yaml (hunk 11 above)
Comment (philei-tt) [reply tt-train/configs/training_configs/training_shakespeare_nanogpt.yaml]: This breaks cpp training's config loading btw

**Comment by philei-tt** [comment]
Seems like python training is a bit slower... I wonder if that's because of bindings overhead. I don't think they should have such a big impact, I would imagine 1-2ms max. 
Also, why there is a slight difference in loss? It seems like it converges to a similar number, but maybe weights initialization is different? What might cause this?
```
py nanogpt 135 steps
Step: 10, Loss: 3.406250, Time: 243.26 ms
Step: 35, Loss: 3.250000, Time: 240.93 ms
Step: 70, Loss: 2.828125, Time: 243.12 ms
Step: 100, Loss: 2.687500, Time: 242.86 ms
Step: 133, Loss: 2.640625, Time: 243.23 ms
Step: 134, Loss: 2.609375, Time: 241.30 ms
Step: 135, Loss: 2.578125, Time: 242.96 ms

cpp nanogpt 135 steps
Step: 10, Loss: 3.34375
Step: 35, Loss: 2.984375
Step: 70, Loss: 2.703125
Step: 100, Loss: 2.640625
Full step time 210.17 ms, cache entries: 79
Step: 133, Loss: 2.625
Full step time 209.827 ms, cache entries: 79
Step: 134, Loss: 2.640625
Full step time 210.24 ms, cache entries: 79
Step: 135, Loss: 2.640625
```

**Comment by philei-tt** [comment]
@athompsonTT I really had to struggle to run py nanogpt. Cursor had to make changes to 9 files and It had to get iteratives fixes like 5 times to get it working.  There was something about wrong loading order and circular imports. Also it added a lot of "from __future__ import annotations". Is it possible you're using newer python version?

File: tt-train/sources/ttml/ttml/models/nanogpt/__init__.py [hunk 16]
```diff
@@ -0,0 +1,173 @@
...
+        self.wte = Embedding(config.vocab_size, config.n_embd)
+        self.wpe = Embedding(config.block_size, config.n_embd)
+
+        # Transformer blocks
+        self.blocks = []
+        for i in range(config.n_layer):
+            block = GPTBlock(config.n_embd, config.n_head, config.dropout, config.bias)
+            self.blocks.append(block)
+            # Register as submodule for parameter tracking
+            setattr(self, f"block_{i}", block)
```
Comment (philei-tt) [inline tt-train/sources/ttml/ttml/models/nanogpt/__init__.py]: This looks kind of ugly. Is the same trick also used when training pytorch model?

Comment (philei-tt) [reply tt-train/sources/ttml/ttml/models/nanogpt/__init__.py]: I checked, that's why torch has nn.ModuleList. Do you think we can add something like that too?

File: tt-train/sources/ttml/ops/reshape_op.cpp (hunk 10 above)
Comment (athompsonTT) [reply tt-train/sources/ttml/ops/reshape_op.cpp:15]: I thought it was necessary for gradient tracking during training

File: tt-train/configs/training_configs/training_shakespeare_nanogpt.yaml (hunk 11 above)
Comment (athompsonTT) [reply tt-train/configs/training_configs/training_shakespeare_nanogpt.yaml]: I'll adjust the python loaders to behave like cpp and undo this change

File: tt-train/sources/ttml/ttml/models/nanogpt/embedding.py (hunk 12 above)
Comment (athompsonTT) [reply tt-train/sources/ttml/ttml/models/nanogpt/embedding.py]: i'll remove the suffix

File: tt-train/sources/ttml/ttml/modules/module_base.py (hunk 13 above)
Comment (athompsonTT) [reply tt-train/sources/ttml/ttml/modules/module_base.py:28]: Correct

**Comment by athompsonTT** [comment]
> @athompsonTT I really had to struggle to run py nanogpt. Cursor had to make changes to 9 files and It had to get iteratives fixes like 5 times to get it working. There was something about wrong loading order and circular imports. Also it added a lot of "from **future** import annotations". Is it possible you're using newer python version?

I'm using Python 3.10.12

File: tt-train/sources/ttml/ttml/models/nanogpt/__init__.py (hunk 16 above)
Comment (athompsonTT) [reply tt-train/sources/ttml/ttml/models/nanogpt/__init__.py]: I'll add ModuleList/ModuleDict

**Review by ayerofieiev-tt — APPROVED** [review]
